apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/ClientTracingFilterInterceptor.java

```diff
@@ -0,0 +1,107 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.finagle;
+
+import com.twitter.util.Future;
+import com.twitter.util.FutureEventListener;
+import org.apache.skywalking.apm.agent.core.context.ContextCarrier;
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.agent.core.context.trace.SpanLayer;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.EnhancedInstance;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.MethodInterceptResult;
+
+import java.lang.reflect.Method;
+
+import static org.apache.skywalking.apm.network.trace.component.ComponentsDefine.FINAGLE;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getLocalContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getMarshalledContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.FinagleCtxs.SW_SPAN;
+
+public class ClientTracingFilterInterceptor extends AbstractInterceptor {
+
+    private static class CacheObjects {
+        private ContextHolder marshlledContextHolder;
+        private ContextHolder localContextHolder;
+
+        private CacheObjects(ContextHolder marshlledContextHolder, ContextHolder localContextHolder) {
+            this.marshlledContextHolder = marshlledContextHolder;
+            this.localContextHolder = localContextHolder;
+        }
+    }
+
+    @Override
+    protected void onConstructImpl(EnhancedInstance objInst, Object[] allArguments) {
+
+    }
+
+    @Override
+    public void beforeMethodImpl(EnhancedInstance enhancedInstance, Method method, Object[] objects, Class<?>[] classes,
+                                 MethodInterceptResult methodInterceptResult) throws Throwable {
+        ContextCarrier contextCarrier = new ContextCarrier();
+        /*
+         * At this time, we can't know the operation name and peer address, so we just use placeholders here, the
+         * operation name will be filled by {@link AnnotationInterceptor$Rpc} and the peer address will be filled by
+         * {@link ClientDestTracingFilterInterceptor} later.
+         */
+        AbstractSpan finagleSpan = ContextManager.createExitSpan("pending", contextCarrier, "");
```

Comment:
Why don't you create the exit span in the `ClientDestTracingFilterInterceptor` directly? ContextCarrier require the peer information, otherwise the topology will break.

Reply:
Because `ClientDestTracingFilter ` may be executed in another thread, not the thread that initiated the rpc request.   It is hard to know when and where the thread change occurs.

Reply:
The peer infomation is added to ContextCarrier in `ClientDestTracingFilterInterceptor` by `ContextCarrierHelper`.

Reply:
So far it is necessary.
https://github.com/apache/skywalking/pull/4441#discussion_r389333225